sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala

```diff
@@ -149,3 +149,73 @@ case class TypedAggregateExpression(
 
   override def nodeName: String = aggregator.getClass.getSimpleName
 }
+
+case class TypedAggregateExpression2(
+    aggregator: Aggregator[Any, Any, Any],
+    inputDeserializer: Expression,
+    inputAttributes: Seq[Attribute],
+    bufferSerializer: Seq[NamedExpression],
+    bufferDeserializer: Expression,
+    bufferClass: Class[_],
+    resultSerializer: Seq[NamedExpression],
+    dataType: DataType) extends DeclarativeAggregate with NonSQLExpression {
+
+  override def nullable: Boolean = true
+
+  override def children: Seq[Expression] = Nil
+
+  override def inputTypes: Seq[AbstractDataType] = Nil
+
+  private val aggregatorLiteral = Literal.fromObject(aggregator)
+
+  private val bufferType = ObjectType(bufferClass)
+
+  override lazy val aggBufferAttributes: Seq[Attribute] = bufferSerializer.map(_.toAttribute)
+
+  override lazy val initialValues: Seq[Expression] = {
+    val zero = Literal.create(aggregator.zero, bufferType)
+    bufferSerializer.map(_ transform {
+      case b: BoundReference => zero
+    })
+  }
+
+  override lazy val updateExpressions: Seq[Expression] = {
+    val reduced = Invoke(
+      aggregatorLiteral,
+      "reduce",
+      bufferType,
+      bufferDeserializer :: inputDeserializer :: Nil)
+
+    bufferSerializer.map(_ transform {
+      case b: BoundReference => reduced
```

Comment:
yes, it's the same problem.

turning an object to a struct in one shot doesn't solve the problem, as we need to flatten it to `Seq[Expression]` here, which will be a list of `GetStructField` and the "one shot" expression still appears many times.

Actually I think we don't need to do anything here. Although the `reduced` expression appears many times, they are same references and subexpression elimination should always work. Or we can introduce the Holder and Reference to make it more clear.